File: backend/app/services/feature_explanation_service.py
import re
from typing import Dict, Any, List
from collections import Counter


class FeatureExplanationService:
    """
    Service responsible for computing feature metrics and evidence explanations
    to make the classifier decisions understandable to the user.
    """

    def extract_features(self, text: str, pattern_evidence: Dict[str, List[str]]) -> Dict[str, Any]:
        """
        Extracts structural features, term counts, and top terms from the text.
        """
        # Tokenize text
        words = [w for w in re.findall(r"\b\w{3,}\b", text.lower())]
        word_count = len(words)

        # Count occurrences of matched keywords
        role_override_count = 0
        data_exfiltration_count = 0

        for category, matched_keywords in pattern_evidence.items():
            count = len(matched_keywords)
            if category in ["override_instructions", "jailbreak_attempt"]:
                role_override_count += count
            elif category == "data_exfiltration":
                data_exfiltration_count += count

        # Compute instruction density (proportion of suspicious words/phrases relative to document length)
        # Approximate: total words matched / total word count
        instruction_density = 0.0
        if word_count > 0:
            # Let's count approximate word matches from keywords
            matched_words_total = 0
            for category, matched_keywords in pattern_evidence.items():
                for kw in matched_keywords:
                    matched_words_total += len(kw.split(" "))
            instruction_density = min(1.0, matched_words_total / word_count)

        # Extract top terms (excluding basic stopwords)
        stop_words = {
            "the", "and", "a", "of", "to", "is", "in", "that", "this", "it", "you", "for", 
            "was", "on", "are", "as", "with", "his", "they", "i", "at", "be", "this", "have"
        }
        filtered_words = [w for w in words if w not in stop_words and len(w) > 3]
        term_freqs = Counter(filtered_words)
        top_terms = [item[0] for item in term_freqs.most_common(5)]

        return {
            "instruction_density": round(instruction_density, 3),
            "role_override_count": role_override_count,
            "data_exfiltration_count": data_exfiltration_count,
            "semantic_similarity_to_malicious_patterns": 0.0,
            "top_terms": top_terms,
            "embedding_or_vectorizer_used": "rule_based_fallback_terms"
        }


# Export a singleton instance
feature_explanation_service = FeatureExplanationService()

File: backend-node/src/services/featureExplanationService.ts
export interface FeatureEvidence {
  instruction_density: number
  role_override_count: number
  data_exfiltration_count: number
  semantic_similarity_to_malicious_patterns: number
  top_terms: string[]
  embedding_or_vectorizer_used: string
}

const STOP_WORDS = new Set([
  'the', 'and', 'a', 'of', 'to', 'is', 'in', 'that', 'this', 'it', 'you', 'for',
  'was', 'on', 'are', 'as', 'with', 'his', 'they', 'i', 'at', 'be', 'have',
])

function round3(value: number): number {
  return Math.round(value * 1000) / 1000
}

class FeatureExplanationService {
  extractFeatures(text: string, patternEvidence: Record<string, string[]>): FeatureEvidence {
    const words = text.toLowerCase().match(/\b\w{3,}\b/g) ?? []
    const wordCount = words.length

    let roleOverrideCount = 0
    let dataExfiltrationCount = 0

    for (const [category, matchedKeywords] of Object.entries(patternEvidence)) {
      const count = matchedKeywords.length
      if (category === 'override_instructions' || category === 'jailbreak_attempt') {
        roleOverrideCount += count
      } else if (category === 'data_exfiltration') {
        dataExfiltrationCount += count
      }
    }

    let instructionDensity = 0.0
    if (wordCount > 0) {
      let matchedWordsTotal = 0
      for (const matchedKeywords of Object.values(patternEvidence)) {
        for (const kw of matchedKeywords) {
          matchedWordsTotal += kw.split(' ').length
        }
      }
      instructionDensity = Math.min(1.0, matchedWordsTotal / wordCount)
    }

    const filteredWords = words.filter((w) => !STOP_WORDS.has(w) && w.length > 3)
    const freqs = new Map<string, number>()
    for (const w of filteredWords) {
      freqs.set(w, (freqs.get(w) ?? 0) + 1)
    }
    const topTerms = [...freqs.entries()]
      .sort((a, b) => b[1] - a[1])
      .slice(0, 5)
      .map(([term]) => term)

    return {
      instruction_density: round3(instructionDensity),
      role_override_count: roleOverrideCount,
      data_exfiltration_count: dataExfiltrationCount,
      semantic_similarity_to_malicious_patterns: 0.0,
      top_terms: topTerms,
      embedding_or_vectorizer_used: 'rule_based_fallback_terms',
    }
  }
}

export const featureExplanationService = new FeatureExplanationService()
